Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,535 @@
+    }
+
+    /// This method asks the pool for a connection usable by the specified `request`, respecting the specified options.
+    ///
+    /// - parameter request: The request that needs a `Connection`
+    /// - parameter preference: The `EventLoopPreference` the connection pool will respect to lease a new connection
+    /// - parameter deadline: The connection timeout
+    /// - Returns: A connection  corresponding to the specified parameters
+    ///
+    /// When the pool is asked for a new connection, it creates a `Key` from the url associated to the `request`. This key
+    /// is used to determine if there already exists an associated `HTTP1ConnectionProvider` in `connectionProviders`
+    /// if it does, the connection provider then takes care of leasing a new connection. If a connection provider doesn't exist, it is created.
+    func getConnection(for request: HTTPClient.Request, preference: HTTPClient.EventLoopPreference, on eventLoop: EventLoop, deadline: NIODeadline?) -> EventLoopFuture<Connection> {
+        let key = Key(request)
+
+        let provider: HTTP1ConnectionProvider = self.connectionProvidersLock.withLock {
+            if let existing = self._connectionProviders[key] {
+                existing.stateLock.withLock {
+                    existing.state.pending += 1
+                }
+                return existing
+            } else {
+                let http1Provider = HTTP1ConnectionProvider(key: key, eventLoop: eventLoop, configuration: self.configuration, parentPool: self)
+                self._connectionProviders[key] = http1Provider
+                http1Provider.stateLock.withLock {
+                    http1Provider.state.pending += 1
+                }
+                return http1Provider
+            }
+        }
+
+        return provider.getConnection(preference: preference)
+    }
+
+    func release(_ connection: Connection) {
+        if let connectionProvider = self[connection.key] {
+            connectionProvider.release(connection: connection)
+        }
+    }
+
+    func syncClose(requiresCleanClose: Bool) throws {
+        let connectionProviders = self.connectionProvidersLock.withLock { _connectionProviders.values }
+        for connectionProvider in connectionProviders {
+            try connectionProvider.syncClose(requiresCleanClose: requiresCleanClose)
+        }
+    }
+
+    /// Used by the `ConnectionPool` to index its `HTTP1ConnectionProvider`s
+    ///
+    /// A key is initialized from a `URL`, it uses the components to derive a hashed value
+    /// used by the `connectionProviders` dictionary to allow retreiving and creating
+    /// connection providers associated to a certain request in constant time.
+    struct Key: Hashable {
+        init(_ request: HTTPClient.Request) {
+            switch request.scheme {
+            case "http":
+                self.scheme = .http
+            case "https":
+                self.scheme = .https
+            default:
```

Comment:
I think we just added `"unix"`

Reply:
@weissi Thanks

Reply:
Done, maybe improvable though but I think it does the job

Reply:
Works correctly in my manual tests